Refactoring of CoreSeedNodesRepository + unit tests

tau3 opened the pull request.

tau3 committed:
Moved creation logic into a separate class

tau3 committed:
Moved NodeAddresses into a separate class

tau3 committed:
Fixed compilation errors in CoreSeedNodeRepositoryFactory

tau3 committed:
Extracted default addresses into a separate class

tau3 committed:
Refactored CoreSeedNodesRepository using NodeAddressLookup

tau3 committed:
Refactored some TODOs in NodeAddress and NodeAddressLookup

tau3 committed:
Extracted AbstractImmutableSetDecorator class

tau3 committed:
Implemented unit test for ImmutableSetDecorator

tau3 committed:
Implemented unit test for NodeAddresses

tau3 committed:
Implemented unit test for NodeAddressLookup

tau3 committed:
Bound NodeAddressLookup to all necessary modules

tau3 committed:
Made DefaultNodeAddresses impossible to instantiate

tau3 committed:
Added headers to new classes

tau3 commented on core/src/main/java/io/bisq/core/network/CoreSeedNodesRepository.java:
```diff
@@ -1,58 +1,37 @@
+import static io.bisq.core.network.DefaultNodeAddresses.DEFAULT_LOCALHOST_SEED_NODE_ADDRESSES;
+import static io.bisq.core.network.DefaultNodeAddresses.DEFAULT_TOR_SEED_NODE_ADDRESSES;
+
+@NotThreadSafe
```
@ManfredKarrer is it ok that i marked the class as not thread safe? Actually it had been safe to access `seedNodeAddresses` from multiple threads before; but it hadn't been so for any other field of the class

ManfredKarrer replied:
Yes you can mark it. 

In general most classes are living only in the context of a single user thread. 
GUI app:
```UserThread.setExecutor(Platform::runLater);```
or if not UI app (seed nodes):
```UserThread.setExecutor(Executors.newSingleThreadExecutor());```

There are only a few areas where multi threading is used: 
- Connection (P2P network)
- Tor startup
- Wallet encryption
- Local persistence
- Price request
- Fee request
- DAO/BSQ

From BitcoinJ we map to user thread as well.

Usual strategy is to map the result from the threaded operation to the user thread (via ``` UserThread.execute(() -> ....);``` so the clients can be sure that the callbacks are living in the same thread context. That model removes a lot of the threading complexity and isolates threading only to those areas where it is required.

cbeams commented:
@tau3, I'd like to push some changes as part of my review. Could you please check the following box on the right side of this PR? Thanks.

![image](https://user-images.githubusercontent.com/301810/36388128-2ef58fd4-159b-11e8-8294-81770418ba98.png)

cbeams commented on core/src/main/java/io/bisq/core/network/CoreSeedNodesRepository.java:
```diff
@@ -1,58 +1,37 @@
+    public Set<NodeAddress> getSeedNodeAddresses() {
+        if (seedNodeAddresses == null) {
+            seedNodeAddresses = lookup.resolveNodeAddresses();
+        }
```
Why the lazy initialization here in the setter? Is this an attempt to avoid an expensive operation? Is this actually an optional code path? If it is usually invoked, or if the operation is not terribly expensive, prefer assigning `seedNodeAddresses` directly in the constructor, such that the field can be marked `final`.

tau3 replied:
`NodeAddressLookup` contains some non-trivial logic, so i thought that invoking it in constructor would not be a good idea. 

> prefer assigning seedNodeAddresses directly in the constructor, such that the field can be marked `final`

totally agree, that way is better. But it would lead us to creation of factory of `SeedNodesRepository`. Is it ok?

cbeams committed:
Remove NetworkStressTest to simplify CoreSeedNodesRepository

NetworkStressTest has been ignored for months because it had become
"outdated" and is therefore now dead code. It is the only place that the
CoreSeedNodesRepository#setTorSeedNodeAddresses
and #setLocalhostSeedNodeAddresses methods were called. Removing it
completely allows for removing these methods and marking the underlying
fields in CoreSeedNodesRepository as final.

cbeams committed:
Simplify CoreSeedNodesRepository#isSeedNode stream pipeline

cbeams committed:
Make CoreSeedNodesRepository#seedNodeAddresses final

Also remove the @NotThreadSafe annotation as this class is now stateless
and immutable and therefore naturally threadsafe.

See https://github.com/bisq-network/exchange/pull/1362#discussion_r169143759
for context.

cbeams committed:
Rename SeedNodesRepository => SeedNodeRepository

The idiomatic approach to naming domain-driven design (DDD)-style
repositories is to use the singular form of the object being
encapsulated by the repository, e.g. BookRepository vs. BooksRepository.

Thus far, the repository interfaces and implementations created here in
Bisq have used the latter, plural form, but going forward we should
revert to the norm. To this end, this commit renames SeedNodesRepository
and its single implementation to SeedNodeRepository and
CoreSeedNodeRepository respectively.

cbeams committed:
Rename CoreSeedNodeRepository => DefaultSeedNodeRepository

The original name was intended to distinguish that this implementation
is the implentation that lives within Bisq's 'core' module. This commit
renames the class to DefaultSeedNodeRepository as this is more idiomatic
and intention-revealing. This implementation is in fact the only one
ever gets used; it is the 'default' (and only) implementation.

cbeams committed:
Rename classes from Node* => SeedNode* as appropriate

Each of the classes here are in fact specific to seed nodes, so have
been renamed accordingly, e.g. from NodeAddresses => SeedNodeAddresses.

cbeams committed:
Favor public visibility for non-private members of package-private types

Public members of package-private classes are in effect still
package-private. Leaving the public modifier in place on these members
(as opposed to marking all members package-private) is beneficial for
two reasons:

 1. It leaves these members looking "normal" to the trained Java eye.
Readers are not left wondering why constructors and methods are
package-private. They can notice, if and when appropriate, that the
class itself is package-private, and can understand the transitive
visibility of each member on that basis.  Otherwise, the class is left
looking conventional, unsurprising, and therefore as easy as possible to
read and understand.

 2. It leaves the visibility of these members controlled by a single
modifier at the class level as opposed to N+1 modifiers (one for every
member + the class itself). Should the class in question for some reason
need to be promoted to public visibility, for example if it were to be
moved to a different package, its members would automatically inherit
the now-public visibility of the class, which is likely to be the
desired state, because if the class needs to be public, at least some of
its members will need to be public too. The maintainer making that
change can (and should) reduce the visibility of any members that should
remain package-private (e.g. those that exist only for for testing
purposes), but in doing so, they make an explicit and self-documenting
decision that will show up at the level of a Git diff.

cbeams committed:
Sort static factory methods at bottom

cbeams committed:
Move io.bisq.core.network.{* => .p2p.seed.*}

Prior to this change, the contents of io.bisq.core.network were in fact
all seednode-related, and the main public type in this package,
DefaultSeedNodeRepository, is an implementation of
io.bisq.network.p2p.seed.SeedNodeRepository. This repackaging makes
better reflects these facts, making it more obvious to the unitiated
what this package is all about, and preserving the unqualified
io.bisq.core.network package for future expansion into non-seednode
related areas.

cbeams committed:
Add @Override where appropriate

cbeams committed:
Indent multiline constructor args

The convention throughout most of the Bisq codebase is this kind of
stacked alignment for multiline constructor arguments.

cbeams reviewed (changes requested):
NACK.

@tau3, I have a number of commits I'd like to push to this PR branch as part of my review. Please check the box that allows maintainers to push commits, thanks.

Once I'm able to push those changes, I'll add a few more comments.

In the meantime, there is a bug somewhere in commits 3e13a0f2262a2f1ca4182a299099edcee4c4f83e and 9c2f0d5c00fdeac377c3a1ac41cca95f1978fd82 that causes the following error on attempting to start Bisq:

```
java.lang.IllegalArgumentException: fullAddress must contain ':'
```

Just try to run `BisqAppMain` from within IDEA, or run `mvn package` followed by `java -jar gui/target/shaded.jar` and you'll see the error get raised immediately as Guice tries and fails to wire up the application.

tau3 committed:
Fixed factory method for NodeAddresses in case of empty string

tau3 commented:
@cbeams yep, there was a silly mistake 😕 
fixed and added unit test.

cbeams commented:
> @cbeams yep, there was a silly mistake 😕
> fixed and added unit test.

Thanks. I've confirmed locally. **Can you please check the "Allow edits from maintainers" option** as requested above in https://github.com/bisq-network/exchange/pull/1362#issuecomment-366744626

![image](https://user-images.githubusercontent.com/301810/36473662-84d7b638-16f5-11e8-81e1-7992740b19fd.png)

I have 11 additional commits as part of my review, and I cannot push them until you've checked this box.

```
~/Work/bisq-network/exchange[tau3-refactoring/core-seed-nodes-repository]
$ git push tau3 tau3-refactoring/core-seed-nodes-repository
fatal: remote error:
  You can't push to git://github.com/tau3/exchange.git
  Use https://github.com/tau3/exchange.git
```

tau3 commented:
@cbeams it is already checked ¯\\_(ツ)_/¯

cbeams reviewed (approved):
ACK.

@tau3, please review each of my additional commits, including the extended commit comments where they exist.

There is more I would have done here, or requested that you do, but I've held off on it, because I have a larger refactoring around seed nodes in mind anyway. Would you be up for having a conversation about that?

In general, your changes are great. Thank you for putting this PR together!

In the future, please tweak the grammar of your commit comments so they adhere to the guidelines at https://chris.beams.io/posts/git-commit#seven-rules. You're already doing most of these; it's just the "imperative mood" one I'm focusing on here.

There are some other nits, like the fact that I don't think we need immutable sets here, and therefore don't need the `ImmutableSetDecorator` class you've introduced, but again, I'm happy to leave this as-is in service of moving on to larger, higher-impact changes. Don't get me wrong here, though: immutable sets are great, and ideally, all collections would be immutable and persistent by default in Java, but they're not, and therefore introducing them always carries certain costs, like the extra class we see here. Generally, if there's not a requirement for collection immutability, just go with the standard collections; it's the convention throughout most of the rest of Bisq anyway.

cbeams commented:
> @cbeams it is already checked ¯\(ツ)/¯

Regarding this: it was an issue with the way I was attempting to push to your remote. My bad; thanks for taking care of this when you originally did. That checkbox will, I believe, now stay checked for future PRs in this repository (thanks to a cookie), but please double-check the next couple PRs to be sure. Thanks.

cbeams committed:
Merge branch 'master' into tau3-refactoring/core-seed-nodes-repository

cbeams committed:
Re-apply "Fix wrong network ID for Dash"

This commit essentially cherry-picks the changes made in ea8cbc1,
re-applying them after the renaming of CoreSeedNodesRepository =>
DefaultSeedNodeRepository that happened earlier on this branch. The
other changes made in that original commit are no longer relevant here,
or have already been made in earlier commits on this refactoring branch.

cbeams commented:
My apologies, @tau3, I never actually merged this PR after our last interactions above. Some merge conflicts had cropped up in the meantime, and I've fixed them. I'll be merging shortly, as soon as Travis gives the OK.
